internal/controller/standalone_pgadmin/users.go

```diff
@@ -0,0 +1,286 @@
+				pgadminVersion,
+			))
+		}
+		return err
+	}
+
+	return r.writePGAdminUsers(ctx, pgadmin, podExecutor)
+}
+
+// reconcilePGAdminMajorVersion execs into the pgAdmin pod and retrieves the pgAdmin major version
+func (r *PGAdminReconciler) reconcilePGAdminMajorVersion(ctx context.Context, exec Executor) (int, error) {
+	script := `
+PGADMIN_DIR=/usr/local/lib/python3.11/site-packages/pgadmin4
+cd $PGADMIN_DIR && python3 -c "import config; print(config.APP_RELEASE)"
+`
+
+	var err error
+	var stdin, stdout, stderr bytes.Buffer
+
+	err = exec(ctx, &stdin, &stdout, &stderr,
+		[]string{"bash", "-ceu", "--", script}...)
+
+	if err != nil {
+		return 0, err
+	}
+
+	return strconv.Atoi(strings.TrimSpace(stdout.String()))
+}
+
+// writePGAdminUsers takes the users in the pgAdmin spec and writes (adds or updates) their data
+// to both pgAdmin and the users.json file that is stored in the pgAdmin secret. If a user is
+// removed from the spec, its data is removed from users.json, but it is not deleted from pgAdmin.
+func (r *PGAdminReconciler) writePGAdminUsers(ctx context.Context, pgadmin *v1beta1.PGAdmin,
+	exec Executor) error {
+	log := logging.FromContext(ctx)
+
+	existingUserSecret := &corev1.Secret{ObjectMeta: naming.StandalonePGAdmin(pgadmin)}
+	err := errors.WithStack(
+		r.Client.Get(ctx, client.ObjectKeyFromObject(existingUserSecret), existingUserSecret))
+	if client.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	intentUserSecret := &corev1.Secret{ObjectMeta: naming.StandalonePGAdmin(pgadmin)}
+	intentUserSecret.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("Secret"))
+
+	intentUserSecret.Annotations = naming.Merge(
+		pgadmin.Spec.Metadata.GetAnnotationsOrNil(),
+	)
+	intentUserSecret.Labels = naming.Merge(
+		pgadmin.Spec.Metadata.GetLabelsOrNil(),
+		map[string]string{
+			naming.LabelStandalonePGAdmin: pgadmin.Name,
+			naming.LabelRole:              naming.RolePGAdmin,
+		})
+
+	// Copy existing data if it exists, else initialize maps
+	if existingUserSecret.Data != nil {
+		intentUserSecret.Data = existingUserSecret.Data
+	} else {
+		intentUserSecret.Data = make(map[string][]byte)
+	}
+	if existingUserSecret.StringData != nil {
+		intentUserSecret.StringData = existingUserSecret.StringData
+	} else {
+		intentUserSecret.StringData = make(map[string]string)
+	}
+
+	setupScript := `
+PGADMIN_DIR=/usr/local/lib/python3.11/site-packages/pgadmin4
+cd $PGADMIN_DIR
+`
+
+	var existingUsersArr []secretPGAdminUser
+	if existingUserSecret.Data["users.json"] != nil {
+		err := json.Unmarshal(existingUserSecret.Data["users.json"], &existingUsersArr)
+		if err != nil {
+			return err
+		}
+	}
+	existingUsersMap := make(map[string]secretPGAdminUser)
+	for _, user := range existingUsersArr {
+		existingUsersMap[user.Username] = user
+	}
+	intentUsers := []secretPGAdminUser{}
+	for _, user := range pgadmin.Spec.Users {
+		var stdin, stdout, stderr bytes.Buffer
+		typeFlag := "--nonadmin"
```

Comment:
Hmm... Why not?

Reply:
It obviously works the same way whether we pass `--nonadmin` or not... If they ever change the way the default behavior works, e.g.:
1. no flag results in an error 
2. no flag defaults to Admin

... then the behavior of our API stays consistent. The only way things would break for us is if they remove `--nonadmin` as an option... :thinking:

Reply:
This seems like a really minor detail to fret over IMO. If pgAdmin changes the way setup.py works in a future version, we will adjust accordingly. We have no way of knowing what their future change will be.

Reply:
I do agree that adding contract tests to ensure setup.py does what we think it does is a good idea though.